mqtt-protocol/src/jsMain/kotlin/it/unibo/pulvreakt/mqtt/MqttProtocol.kt

```diff
@@ -1,47 +1,146 @@
 package it.unibo.pulvreakt.mqtt
 
 import arrow.core.Either
+import arrow.core.raise.either
+import arrow.core.raise.ensureNotNull
+import arrow.core.right
+import io.github.oshai.kotlinlogging.KotlinLogging
 import it.unibo.pulvreakt.api.communication.protocol.Entity
 import it.unibo.pulvreakt.api.communication.protocol.Protocol
 import it.unibo.pulvreakt.errors.protocol.ProtocolError
 import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.asSharedFlow
 import org.kodein.di.DI
+import kotlin.random.Random
 
 /**
- * Represents the MQTT protocol used in PulvReAKt.
+ * MQTT Protocol implementation on JS side.
  */
-@Suppress("UnusedPrivateProperty")
 actual class MqttProtocol actual constructor(
     private val host: String,
     private val port: Int,
     private val username: String?,
     private val password: String?,
-    private val coroutineDispatcher: CoroutineDispatcher,
+    coroutineDispatcher: CoroutineDispatcher,
 ) : Protocol {
-    override suspend fun setupChannel(source: Entity, destination: Entity) {
-        TODO("Not yet implemented")
-    }
+    private val mainTopic = "PulvReAKt"
+
+    override lateinit var di: DI
+    private val logger = KotlinLogging.logger("MqttProtocol")
+
+    private val registeredTopics = mutableMapOf<Pair<Entity, Entity>, String>()
+    private val topicChannels = mutableMapOf<String, MutableSharedFlow<ByteArray>>()
+    private lateinit var client: MqttClient
 
-    override suspend fun writeToChannel(from: Entity, to: Entity, message: ByteArray): Either<ProtocolError, Unit> {
-        TODO("Not yet implemented")
+    override suspend fun setupChannel(
+        source: Entity,
+        destination: Entity
+    ) {
+        logger.debug { "-Setting up channel for entity $source to $destination" }
+        registeredTopics += (source to destination) to toTopics(source, destination)
+        registeredTopics += (destination to source) to toTopics(destination, source)
+        topicChannels += toTopics(source, destination) to MutableSharedFlow(1)
+        topicChannels += toTopics(destination, source) to MutableSharedFlow(1)
     }
 
-    override fun readFromChannel(from: Entity, to: Entity): Either<ProtocolError, Flow<ByteArray>> {
-        TODO("Not yet implemented")
+    override suspend fun writeToChannel(
+        from: Entity,
+        to: Entity,
+        message: ByteArray
+    ): Either<ProtocolError, Unit> = either {
+        val topic = registeredTopics[Pair(from, to)]
+        logger.debug { "-Writing message $message to topic $topic" }
+
+        ensureNotNull(topic) { ProtocolError.EntityNotRegistered(to) }
+
+        Either.catch {
+            client.publish(
+                topic,
+                message.decodeToString(),
+                options = js("{ qos: 2, retain: true }")
```

Comment:
Instead of hardcoding these parameters, what do you think if we pass them via the class constructor? Perhaps with default parameters like `qos = 2` and `retain = true` so that the user can override them.
Could you make this update?
Of course this change should be reflected also in the common code